orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/ExecutionRunnerSupport.java

```diff
@@ -37,99 +45,154 @@ public ExecutionRunnerSupport(Collection<StageDefinitionBuilder> stageDefinition
+  private <T extends Execution<T>> void injectStage(
+    Stage<T> parent,
+    Stage<T> stage,
+    SyntheticStageOwner type
+  ) {
+    List<Stage<T>> stages = parent.getExecution().getStages();
+    int index = stages.indexOf(parent);
+    int offset = type == STAGE_BEFORE ? 0 : 1;
+    stages.add(index + offset, stage);
+    stage.setParentStageId(parent.getId());
+  }
+
+  // TODO: change callback type to Consumer<TaskDefinition>
+  protected <T extends Execution<T>> void planTasks(Stage<T> stage, TaskNode.TaskGraph taskGraph, boolean isSubGraph, Consumer<com.netflix.spinnaker.orca.pipeline.model.Task> callback) {
     for (ListIterator<TaskNode> itr = taskGraph.listIterator(); itr.hasNext(); ) {
       boolean isStart = !itr.hasPrevious();
       // do this after calling itr.hasPrevious because ListIterator is stupid
       TaskNode taskDef = itr.next();
       boolean isEnd = !itr.hasNext();
 
       if (taskDef instanceof TaskDefinition) {
-        DefaultTask task = new DefaultTask();
-        if (isStart) {
-          if (isSubGraph) {
-            task.setLoopStart(true);
-          } else {
-            task.setStageStart(true);
-          }
-        }
-        task.setId(String.valueOf((stage.getTasks().size() + 1)));
-        task.setName(((TaskDefinition) taskDef).getName());
-        task.setStatus(NOT_STARTED);
-        task.setImplementingClass(((TaskDefinition) taskDef).getImplementingClass());
-        if (isEnd) {
-          if (isSubGraph) {
-            task.setLoopEnd(true);
-          } else {
-            // TODO: may need to revisit this for parallel stages like bake & deploy
-            task.setStageEnd(true);
-          }
-        }
-        stage.getTasks().add(task);
+        planTask(stage, (TaskDefinition) taskDef, isSubGraph, isStart, isEnd, callback);
       } else if (taskDef instanceof TaskNode.TaskGraph) {
-        planTasks(stage, (TaskNode.TaskGraph) taskDef, true);
+        planTasks(stage, (TaskNode.TaskGraph) taskDef, true, callback);
       } else {
         throw new UnsupportedOperationException(format("Unknown TaskNode type %s", taskDef.getClass().getName()));
       }
     }
   }
 
+  private <T extends Execution<T>> void planTask(Stage<T> stage, TaskDefinition taskDef, boolean isSubGraph, boolean isStart, boolean isEnd, Consumer<com.netflix.spinnaker.orca.pipeline.model.Task> callback) {
+    DefaultTask task = new DefaultTask();
+    if (isStart) {
+      if (isSubGraph) {
+        task.setLoopStart(true);
+      } else {
+        task.setStageStart(true);
+      }
+    }
+    task.setId(String.valueOf((stage.getTasks().size() + 1)));
+    task.setName(taskDef.getName());
+    task.setStatus(NOT_STARTED);
+    task.setImplementingClass(taskDef.getImplementingClass());
+    if (isEnd) {
+      if (isSubGraph) {
+        task.setLoopEnd(true);
+      } else {
+        // TODO: may need to revisit this for parallel stages like bake & deploy
```

Comment:
I tried it out and yeah, we still need it. The issue is closed but I think we need a dependency rev to get the fix.